Poetry build system

mquinnfd opened the pull request:
## Switching to the Poetry build system

> [!TIP]
> This PR was worked on iteratively in `Draft` form, so there's a bit of history in the comments here.
>
> This is being suggested for integration to make building and bundling the front end a bit easier, comments welcome particularly if you have strong opinions on `poetry`

### In a Nutshell

This PR changes the build system used to package Locust over to [Poetry](https://python-poetry.org/).

The main use-case for this currently, was to allow the removal of the (snazzy new) dynamically generated front end bundle from the actual source code, and just let the build process create these for the package during the build.

#### Caveat

<details><summary>Details</summary>
<br>
<p>
I've tried to keep the package output and metadata like-for-like, because I havn't looked behind the curtain at the Locust release process, CI runners, etc. - it's likely this may need to change for that purpose.

There's plenty I am blissfully unaware of when it comes to maintaining this project, this is just a jumping off point for the sake of conversation - I have no particular love of Poetry, though I do use this day-to-day for projects currently. Alternatives noted were [PDM](https://github.com/pdm-project/pdm) (and now [Rye](https://rye-up.com/) by "the Ruff people") but I have no idea if these support the king of things we're doing here currently.
</p>
</details> 

---

### What

Adds the Poetry build system to the Locust package, this changes the build backend to Poetry's [PEP-517](https://peps.python.org/pep-0517/) compliant build back-end; [Poetry Core](https://github.com/python-poetry/poetry-core)

- Added Poetry build system
- Don't check in Poetry lock file, as we're often used as a library (worth discussion) - but do include it in generated builds, for provenance
- Poetry plugin [poetry-dynamic-versioning](https://github.com/mtkennerly/poetry-dynamic-versioning) does the job of [setuptools-scm](https://github.com/pypa/setuptools_scm/) to keep the version number in line with VCS info

### Why

Discussed with @cyberw here: https://github.com/locustio/locust/pull/2405#issuecomment-1806527834

- Allows for pre-build script, which allows us to build the front end on-demand and remove it from VCS
- While Poetry is far from perfect, it is at least [well maintained](https://github.com/python-poetry/poetry) by a large community
- Package size reduced by 10% due to not packaging tests *etc.* in the wheel (*were these necessary somehow? library users maybe?*)

---

### Try it Out

*Local dev installation*
<details><summary>Details</summary>
<br>
<code>make install</code>
<br><br>
<ul>
  <li>Installs project dependencies</li>
  <li>Installs editable local package of `locust`</li>
</ul>
</details> 

*Packaging*

<details><summary>Details</summary>
<br>
<code>make build</code>
<br><br>
<ul>
  <li>Installs `poetry` and the required plugin</li>
  <li>Builds the project, including the web UI</li>
  <li>Packages this, and homogenises the wheel package name</li>
</ul>
</details> 

### TODO

I wanted to put there out there for discussion before doing things like:
- Rework the Github actions workflows for the CI build
- Adding documentation and updating the `README`
- Adding _improvements_ to the build process

mquinnfd commented:
For the consideration of @cyberw and community, apologies on the delay, sometimes life happens 😓

mquinnfd commented on Makefile:
```diff
@@ -15,9 +22,3 @@ build_docs:
 changelog:
 	@echo "Not supported any more. Run ./generate_changelog.py <version_number> instead!"
-
-sass_watch:
```
Pretty sure these are now unused, figured it would be OK to yank them out

andrewbaldwin44 replied:
Yes you're right, sorry I missed those when removing the legacy UI :+1:

mquinnfd commented on pyproject.toml:
```diff
@@ -1,28 +1,20 @@
+version = "0.0.0"
+license = "MIT"
+readme = "README.md"
+authors = ["Jonatan Heyman", "Carl Byström", "Joakim Hamrén", "Hugo Heyman"]
```
Just took this info from the main site 😓

cyberw replied:
Make it 

`authors = ["Jonatan Heyman", "Lars Holmberg", "Andrew Baldwin"]`

That's the most accurate without listing every contributor :)

mquinnfd commented on pyproject.toml:
```diff
@@ -41,22 +33,81 @@ classifiers = [
+
+[tool.poetry-dynamic-versioning.files."locust/_version.py"]
+persistent-substitution = true
+initial-content = """
```
Works with the current setup like-for-like, otherwise could be tweaked if the file can change

mquinnfd commented:
I'm also unsure if you have squash-on-merge turned on for the repo, just an FYI as I have not lovingly rebased my commits currently

cyberw commented:
First run of make install failed for me. Probably the shell doesnt update its list of installed things in the middle of a && (could be enough to move `poetry self add...` to a new line in the Makefile)
```
...
Using cached jaraco.classes-3.4.0-py3-none-any.whl (6.8 kB)
Installing collected packages: jaraco.classes, cleo, build, keyring, cachecontrol, poetry-plugin-export, poetry
Successfully installed build-1.2.1 cachecontrol-0.14.0 cleo-2.1.0 jaraco.classes-3.4.0 keyring-24.3.1 poetry-1.8.3 poetry-plugin-export-1.8.0
/bin/sh: poetry: command not found
make: *** [setup_build] Error 127
```

cyberw commented:
Perhaps it makes more sense to just assume poetry is installed (possibly giving a slightly nicer error message if it is not)

cyberw commented:
Make build fails on yarn not being installed too. Bonus points if we can give a nice error message or even provide an action to install that.

cyberw commented:
Make install doesnt actually add locust to my path. Do i need to do something more to set up poetry correctly? (`python3 -m locust ...` works though)

cyberw commented:
I think the versioning is only *almost* right. 

```
python -m locust -V
locust 2.28.0 from /Users/lars/git/locust/locust (Python 3.12.2, OpenSSL 3.3.0)
```

On an un-tagged master it should be something like locust-2.28.1.dev5

mquinnfd commented:
> First run of make install failed for me. Probably the shell doesnt update its list of installed things in the middle of a && (could be enough to move `poetry self add...` to a new line in the Makefile)
> 
> ```
> ...
> Using cached jaraco.classes-3.4.0-py3-none-any.whl (6.8 kB)
> Installing collected packages: jaraco.classes, cleo, build, keyring, cachecontrol, poetry-plugin-export, poetry
> Successfully installed build-1.2.1 cachecontrol-0.14.0 cleo-2.1.0 jaraco.classes-3.4.0 keyring-24.3.1 poetry-1.8.3 poetry-plugin-export-1.8.0
> /bin/sh: poetry: command not found
> make: *** [setup_build] Error 127
> ```

Great shout! I ran this again in a clean `venv` and fixed with https://github.com/locustio/locust/pull/2725/commits/3b5540546505dfff97cc7e455099387e3c9dcd5a

mquinnfd commented:
> Make build fails on yarn not being installed too. Bonus points if we can give a nice error message or even provide an action to install that.

Yeah good idea, I don't think there's a reliable cross-platform way to do this really - I've added https://github.com/locustio/locust/pull/2725/commits/9c90aec91644d68cbaed96b97a6c2d2c6f1bd79c which gives the following behaviour when you don't have the `yarn` binary available:

```
<username>@<hostname> locust-fork % make build
Locust requires the yarn binary to be available in this shell to build the web front-end.
See: https://classic.yarnpkg.com/lang/en/docs/install
make: *** [check-yarn] Error 1
```

We could certainly do something like attempting to install it via NPM after checking if NPM is there?
Same goes with `poetry` - whatever is easier for people, could just warn out to the terminal alone.

mquinnfd commented:
> Make install doesnt actually add locust to my path. Do i need to do something more to set up poetry correctly? (`python3 -m locust ...` works though)

Would you mind trying again on a clean `venv` if possible? (with the branch up to date)
I added https://github.com/locustio/locust/pull/2725/commits/7e290ff1007f0a4e984e6d2d15f5483a5c2cce4e which, for me, results in:

- `make install`
- `poetry` is on the path, and can be called

mquinnfd commented:
> I think the versioning is only _almost_ right.
> 
> ```
> python -m locust -V
> locust 2.28.0 from /Users/lars/git/locust/locust (Python 3.12.2, OpenSSL 3.3.0)
> ```
> 
> On an un-tagged master it should be something like locust-2.28.1.dev5 (this version is needed to make good versions for pre-releases that are made whenever something is merged to master, see https://pypi.org/project/locust/#history)

I'll need to take a look at this behaviour - it could be that I've made a load of tags testing this and something's getting a bit confused, I can try on a clean branch maybe - it should reflect your latest local git tag though right? What is that, in your case?

```
user@computer % python -m locust -V
locust 2.28.0.dev0 from /Users/user/dev/locust-fork/locust (Python 3.9.16, OpenSSL 1.1.1v)
user@computer % git describe --tags --abbrev=0
2.28.0.dev0
```

cyberw commented:
> > Make build fails on yarn not being installed too. Bonus points if we can give a nice error message or even provide an action to install that.
> 
> Yeah good idea, I don't think there's a reliable cross-platform way to do this really - I've added [9c90aec](https://github.com/locustio/locust/commit/9c90aec91644d68cbaed96b97a6c2d2c6f1bd79c) which gives the following behaviour when you don't have the `yarn` binary available:

I think this is a reasonable compromise. People may want to install node and yarn in various ways so automating that is out of scope I think.

cyberw commented:
> user@computer % python -m locust -V
> locust 2.28.0.dev0 from /Users/user/dev/locust-fork/locust (Python 3.9.16, OpenSSL 1.1.1v)
> user@computer % git describe --tags --abbrev=0
> 2.28.0.dev0

Yea, its weird... I wonder if anything changes if you merge latest master onto your branch? 

```
> git describe --tags --abbrev=0
2.28.0
> locust -V
locust 2.28.0 from /Users/lars/git/locust/locust (Python 3.12.2, OpenSSL 3.3.0)
```

cyberw commented:
> > Make install doesnt actually add locust to my path. Do i need to do something more to set up poetry correctly? (`python3 -m locust ...` works though)
> 
> Would you mind trying again on a clean `venv` if possible? (with the branch up to date) I added [7e290ff](https://github.com/locustio/locust/commit/7e290ff1007f0a4e984e6d2d15f5483a5c2cce4e) which, for me, results in:
> 
> * `make install`
> * `poetry` is on the path, and can be called

Works now!

cyberw commented:
> * If `poetry` not installed, `pip install poetry` etc. -> continue
> * (build happens)
> 
> Same goes with `poetry` - whatever is easier for people, could just warn out to the terminal alone.

Perhaps we should not install poetry either? Poetry docs specifically say never to install poetry in the same venv as your project, which is what we're currently doing, right? https://python-poetry.org/docs/#installation

cyberw commented:
Looking at the rename-wheel.sh script. Poetry claims to have solved the issue? https://github.com/python-poetry/poetry-core/pull/591

I havent looked into it though...

andrewbaldwin44 commented:
> > > Make build fails on yarn not being installed too. Bonus points if we can give a nice error message or even provide an action to install that.
> > 
> > 
> > Yeah good idea, I don't think there's a reliable cross-platform way to do this really - I've added [9c90aec](https://github.com/locustio/locust/commit/9c90aec91644d68cbaed96b97a6c2d2c6f1bd79c) which gives the following behaviour when you don't have the `yarn` binary available:
> 
> I think this is a reasonable compromise. People may want to install node and yarn in various ways so automating that is out of scope I think.

I'm really excited about this change and finally getting the dist out of the VCS! Thanks for working on this @mquinnfd!

I would agree with just showing an error message telling the user yarn is not installed, we could even link to the `developing-locust` file. In the docs I suggest to use nvm and avoid installing yarn thru npm since sometimes the two don't get along. Maybe `developing-locust` could also include some notes on the best way to poetry (like installing it globally and not within venv as @cyberw mentioned)?

andrewbaldwin44 commented on pyproject.toml:
```diff
@@ -41,22 +33,81 @@ classifiers = [
-file = "README.md"
-content-type = "text/markdown"
+[tool.poetry.urls]
+"Help/Questions" = "https://stackoverflow.com/questions/tagged/locust"
```
Would Slack be a more useful link than Stackoverflow for questions? I don't know if any Locust maintainers are active on Stackoverflow

cyberw replied:
I am :) it is linked to the stack-overflow-rss channel on slack. But tbh, I think we should link people to the github discussions forum now.

cyberw replied:
So https://github.com/orgs/locustio/discussions/categories/q-a should be the URL in this case

mquinnfd replied:
Updated this to add them all, pretty sure the number/title of the URLs are arbitrary - we can drop SO and Slack links if you want as well 👍

mquinnfd commented:
> > * If `poetry` not installed, `pip install poetry` etc. -> continue
> > * (build happens)
> > 
> > Same goes with `poetry` - whatever is easier for people, could just warn out to the terminal alone.
> 
> Perhaps we should not install poetry either? Poetry docs specifically say never to install poetry in the same venv as your project, which is what we're currently doing, right? https://python-poetry.org/docs/#installation
> 
> (personally I'd probably keep poetry as a global install, having multiple "levels" of venvs is too much for me :) )

Yeah that works, I've adjusted this behaviour with https://github.com/locustio/locust/pull/2725/commits/1ef12b284675a75bfaf806d7a6b828ce47e675d0 to just moan about `yarn` or `poetry` being missing



> > > > Make build fails on yarn not being installed too. Bonus points if we can give a nice error message or even provide an action to install that.
> > > 
> > > 
> > > Yeah good idea, I don't think there's a reliable cross-platform way to do this really - I've added [9c90aec](https://github.com/locustio/locust/commit/9c90aec91644d68cbaed96b97a6c2d2c6f1bd79c) which gives the following behaviour when you don't have the `yarn` binary available:
> > 
> > 
> > I think this is a reasonable compromise. People may want to install node and yarn in various ways so automating that is out of scope I think.
> 
> I'm really excited about this change and finally getting the dist out of the VCS! Thanks for working on this @mquinnfd!
> 
> I would agree with just showing an error message telling the user yarn is not installed, we could even link to the `developing-locust` file. In the docs I suggest to use nvm and avoid installing yarn thru npm since sometimes the two don't get along. Maybe `developing-locust` could also include some notes on the best way to poetry (like installing it globally and not within venv as @cyberw mentioned)?

As for this, I've changed the links to the `developing-locust` page - definitely easier to manage, obviously if this ends up being a viable addition, I'd update that page for the `poetry` instructions (global vs. local `venv` etc.)

It would be nice to link to the specific heading with `#` within `developing-locust` but this does create a link that needs to be updated in sync with that page, which is not so nice

We could:
- Link to both `poetry` / `yarn` own installation docs *as well* as the `developing-locust` page?
- Just link to `developing-locust`

mquinnfd commented:
> Looking at the rename-wheel.sh script. Poetry claims to have solved the issue? [python-poetry/poetry-core#591](https://github.com/python-poetry/poetry-core/pull/591)
> 
> I havent looked into it though...

So it's the one part of this that's a bit jank - using a build script with `poetry` seems to nominally infer that you're using it to precompile C modules and whanot, so it naturally assumes you'd want to create an architecture-specific package, which we're not fussed about currently.

I've followed this particular rabbit-hole fairly deep, there's a fair amount of threads going on there, the [#591](https://github.com/python-poetry/poetry-core/pull/591) one seems to be partially merged... the useful part of this would have been the regex

It's possible that I've missed some other solution, it's a little frustrating though, but maybe it's something I could PR to `poetry-core` instead like the original poster

I think another *possible* solution to this would be to create a `poetry` plugin to run the pre-build hook instead, which I *think* should be doable, and wouldn't result in this behaviour. It might be fun to look at that anyway if I have time.

cyberw commented:
> It would be nice to link to the specific heading with # within developing-locust but this does create a link that needs to be updated in sync with that page, which is not so nice

I think this is the best option.

cyberw commented:
> So it's the one part of this that's a bit jank

You can leave that as is for now. Sounds like something poetry might fix later on by themselves and as long as it works…

mquinnfd commented:
> > user@computer % python -m locust -V
> > locust 2.28.0.dev0 from /Users/user/dev/locust-fork/locust (Python 3.9.16, OpenSSL 1.1.1v)
> > user@computer % git describe --tags --abbrev=0
> > 2.28.0.dev0
> 
> Yea, its weird... I wonder if anything changes if you merge latest master onto your branch?
> 
> ```
> > git describe --tags --abbrev=0
> 2.28.0
> > locust -V
> locust 2.28.0 from /Users/lars/git/locust/locust (Python 3.12.2, OpenSSL 3.3.0)
> ```

Ahh - I think I understand, it's just the *possible* prefix of `locust-` on the version, I think I can tweak this in the tag-finding regex, will update

mquinnfd commented:
> I think the versioning is only _almost_ right.
> 
> ```
> python -m locust -V
> locust 2.28.0 from /Users/lars/git/locust/locust (Python 3.12.2, OpenSSL 3.3.0)
> ```
> 
> On an un-tagged master it should be something like locust-2.28.1.dev5 (this version is needed to make good versions for pre-releases that are made whenever something is merged to master, see https://pypi.org/project/locust/#history)

Ah - do you mean that this should be looking for tags that are *possibly* prefixed with `locust-` like `locust-2.28.1.dev5`? I believe I can do this by tweaking the regex, I assume this is `PEP440` compliant (I believe `poetry` itself mandates this)

Looking at the history page I only see the `2.28.1.dev25` tags on the versions, which the current state of the branch *should* cover 🤔 - or is it that with `twine` publishing workflow etc. we need this?

mquinnfd commented:
> > It would be nice to link to the specific heading with # within developing-locust but this does create a link that needs to be updated in sync with that page, which is not so nice
> 
> I think this is the best option.

I interpreted this as "put the hash links in" :D - this is in https://github.com/locustio/locust/pull/2725/commits/6e6e9f963473c096115766b116bc0d395d0747a0

If it's the opposite please let me know and I'll revert it

cyberw commented:
> > > It would be nice to link to the specific heading with # within developing-locust but this does create a link that needs to be updated in sync with that page, which is not so nice
> > 
> > 
> > I think this is the best option.
> 
> I interpreted this as "put the hash links in" :D - this is in [6e6e9f9](https://github.com/locustio/locust/commit/6e6e9f963473c096115766b116bc0d395d0747a0)
> 
> If it's the opposite please let me know and I'll revert it

Perfect. Anything else that needs to be done? (except for merging/resolving against master)

cyberw commented:
> > I think the versioning is only _almost_ right.
> > ```
> > python -m locust -V
> > locust 2.28.0 from /Users/lars/git/locust/locust (Python 3.12.2, OpenSSL 3.3.0)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > On an un-tagged master it should be something like locust-2.28.1.dev5 (this version is needed to make good versions for pre-releases that are made whenever something is merged to master, see https://pypi.org/project/locust/#history)
> 
> Ah - do you mean that this should be looking for tags that are _possibly_ prefixed with `locust-` like `locust-2.28.1.dev5`? I believe I can do this by tweaking the regex, I assume this is `PEP440` compliant (I believe `poetry` itself mandates this)
> 
> Looking at the history page I only see the `2.28.1.dev25` tags on the versions, which the current state of the branch _should_ cover 🤔 - or is it that with `twine` publishing workflow etc. we need this?

setuptools_scm generates a version based on the number of commits since the last tag (there are no .devX-tags). Can that be done with poetry-dynamic-versioning too?

mquinnfd commented:
> > > I think the versioning is only _almost_ right.
> > > ```
> > > python -m locust -V
> > > locust 2.28.0 from /Users/lars/git/locust/locust (Python 3.12.2, OpenSSL 3.3.0)
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > On an un-tagged master it should be something like locust-2.28.1.dev5 (this version is needed to make good versions for pre-releases that are made whenever something is merged to master, see https://pypi.org/project/locust/#history)
> > 
> > 
> > Ah - do you mean that this should be looking for tags that are _possibly_ prefixed with `locust-` like `locust-2.28.1.dev5`? I believe I can do this by tweaking the regex, I assume this is `PEP440` compliant (I believe `poetry` itself mandates this)
> > Looking at the history page I only see the `2.28.1.dev25` tags on the versions, which the current state of the branch _should_ cover 🤔 - or is it that with `twine` publishing workflow etc. we need this?
> 
> setuptools_scm generates a version based on the number of commits since the last tag (there are no .devX-tags). Can that be done with poetry-dynamic-versioning too? (my first guess is that it is what is called `distance` there)
> 
> Perhaps this is close: format-jinja = """ {%- if distance == 0 -%} {{ serialize_pep440(base, stage, revision) }} {%- elif revision is not none -%} {{ serialize_pep440(base, stage, revision + 1, dev=distance, metadata=[commit]) }} {%- else -%} {{ serialize_pep440(bump_version(base), stage, revision, dev=distance, metadata=[commit]) }} {%- endif -%} """

Ah - I getcha - we want to generate the these `devX` tags like:
- `2.28.1.dev25` - where the `25` is 25 commits from the last bump?

As you said yeah I think this is the `distance` in that plugin - so I'll take a look at this and see if I can make it do this - what you have looks about right, will update today if I can

mquinnfd commented:
> > > > It would be nice to link to the specific heading with # within developing-locust but this does create a link that needs to be updated in sync with that page, which is not so nice
> > > 
> > > 
> > > I think this is the best option.
> > 
> > 
> > I interpreted this as "put the hash links in" :D - this is in [6e6e9f9](https://github.com/locustio/locust/commit/6e6e9f963473c096115766b116bc0d395d0747a0)
> > If it's the opposite please let me know and I'll revert it
> 
> Perfect. Anything else that needs to be done? (except for merging/resolving against master)

Docs really, and any rework to the Github CI job - I think for the most part these are working given that the regular `pip` install should just work, I'll look at getting these both done

cyberw commented:
The version in the filename does seem slightly different in the logs: locust-0.0.1.dev1+60066b8
instead of: 0.1.dev1 

But maybe that is only on a non-master build on GH. 

Versions buily locally (not on GH) can't be too far off as Locust will (intentionally) refuse to connect master-worker that differ in major revision (and it is messy if I cant make changes in master without deploying workers again all the time :)

mquinnfd commented:
> Nice. Windows looks a bit unstable so I re-ran it...
> 
> docker build seems to have some issues, havent really investigated (this job is skipped on PR builds due to versioning problems, but I think it should work locally). Maybe only needs to change line 19 in Dockerfile to use poetry.
> 
> ```
> BUILDKIT_CONTEXT_KEEP_GIT_DIR=1 docker build .
> ...
> #15 5.036    ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 250.3/250.3 kB 38.4 MB/s eta 0:00:00
> #15 5.055 Building wheels for collected packages: locust, psutil
> #15 5.056   Building wheel for locust (pyproject.toml): started
> #15 5.319   Building wheel for locust (pyproject.toml): finished with status 'error'
> #15 5.322   error: subprocess-exited-with-error
> #15 5.322
> #15 5.322   × Building wheel for locust (pyproject.toml) did not run successfully.
> #15 5.322   │ exit code: 1
> #15 5.322   ╰─> [37 lines of output]
> #15 5.322       Building front end...
> #15 5.322       Traceback (most recent call last):
> #15 5.322         File "/build/build.py", line 10, in <module>
> #15 5.322           build()
> #15 5.322         File "/build/build.py", line 6, in build
> #15 5.322           subprocess.run(["make", "frontend_build"])
> #15 5.322         File "/usr/local/lib/python3.11/subprocess.py", line 548, in run
> #15 5.322           with Popen(*popenargs, **kwargs) as process:
> #15 5.322                ^^^^^^^^^^^^^^^^^^^^^^^^^^^
> #15 5.322         File "/usr/local/lib/python3.11/subprocess.py", line 1026, in __init__
> #15 5.322           self._execute_child(args, executable, preexec_fn, close_fds,
> #15 5.322         File "/usr/local/lib/python3.11/subprocess.py", line 1955, in _execute_child
> #15 5.322           raise child_exception_type(errno_num, err_msg, err_filename)
> #15 5.322       FileNotFoundError: [Errno 2] No such file or directory: 'make'
> #15 5.322       Traceback (most recent call last):
> #15 5.322         File "/opt/venv/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 353, in <module>
> #15 5.322           main()
> #15 5.322         File "/opt/venv/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 335, in main
> #15 5.322           json_out['return_val'] = hook(**hook_input['kwargs'])
> #15 5.322                                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> #15 5.322         File "/opt/venv/lib/python3.11/site-packages/pip/_vendor/pyproject_hooks/_in_process/_in_process.py", line 251, in build_wheel
> #15 5.322           return _build_backend().build_wheel(wheel_directory, config_settings,
> #15 5.322                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> #15 5.322         File "/tmp/pip-build-env-ifae4gfa/overlay/lib/python3.11/site-packages/poetry/core/masonry/api.py", line 58, in build_wheel
> #15 5.322           return WheelBuilder.make_in(
> #15 5.322                  ^^^^^^^^^^^^^^^^^^^^^
> #15 5.322         File "/tmp/pip-build-env-ifae4gfa/overlay/lib/python3.11/site-packages/poetry/core/masonry/builders/wheel.py", line 88, in make_in
> #15 5.322           wb.build(target_dir=directory)
> #15 5.322         File "/tmp/pip-build-env-ifae4gfa/overlay/lib/python3.11/site-packages/poetry/core/masonry/builders/wheel.py", line 123, in build
> #15 5.322           self._build(zip_file)
> #15 5.322         File "/tmp/pip-build-env-ifae4gfa/overlay/lib/python3.11/site-packages/poetry/core/masonry/builders/wheel.py", line 172, in _build
> #15 5.322           self._run_build_script(self._package.build_script)
> #15 5.322         File "/tmp/pip-build-env-ifae4gfa/overlay/lib/python3.11/site-packages/poetry/core/masonry/builders/wheel.py", line 262, in _run_build_script
> #15 5.322           subprocess.check_call([self.executable.as_posix(), build_script])
> #15 5.322         File "/usr/local/lib/python3.11/subprocess.py", line 413, in check_call
> #15 5.322           raise CalledProcessError(retcode, cmd)
> #15 5.322       subprocess.CalledProcessError: Command '['/opt/venv/bin/python', 'build.py']' returned non-zero exit status 1.
> #15 5.322       [end of output]
> #15 5.322
> #15 5.322   note: This error originates from a subprocess, and is likely not a problem with pip.
> #15 5.322   ERROR: Failed building wheel for locust
> #15 5.323   Building wheel for psutil (pyproject.toml): started
> #15 6.049   Building wheel for psutil (pyproject.toml): finished with status 'done'
> #15 6.050   Created wheel for psutil: filename=psutil-5.9.8-cp311-abi3-linux_aarch64.whl size=290020 sha256=c3dc695654c30dd593c3dcdf25af8dc426daf86c4b62e4e9921671dfc0218b4d
> #15 6.050   Stored in directory: /root/.cache/pip/wheels/a6/1e/65/fb0ad37886dca3f25a0aa8e50f4903c5bdbde4bb8a9b1e27de
> #15 6.052 Successfully built psutil
> #15 6.052 Failed to build locust
> #15 6.053 ERROR: Could not build wheels for locust, which is required to install pyproject.toml-based projects
> #15 ERROR: process "/bin/sh -c pip install /build/" did not complete successfully: exit code: 1
> 
> #14 [webui-builder 4/5] RUN yarn webui:install --production --network-timeout 60000
> #14 CANCELED
> ------
>  > [builder 5/5] RUN pip install /build/:
> 5.322
> 5.322   note: This error originates from a subprocess, and is likely not a problem with pip.
> 5.322   ERROR: Failed building wheel for locust
> 5.323   Building wheel for psutil (pyproject.toml): started
> 6.049   Building wheel for psutil (pyproject.toml): finished with status 'done'
> 6.050   Created wheel for psutil: filename=psutil-5.9.8-cp311-abi3-linux_aarch64.whl size=290020 sha256=c3dc695654c30dd593c3dcdf25af8dc426daf86c4b62e4e9921671dfc0218b4d
> 6.050   Stored in directory: /root/.cache/pip/wheels/a6/1e/65/fb0ad37886dca3f25a0aa8e50f4903c5bdbde4bb8a9b1e27de
> 6.052 Successfully built psutil
> 6.052 Failed to build locust
> 6.053 ERROR: Could not build wheels for locust, which is required to install pyproject.toml-based projects
> ------
> Dockerfile:19
> --------------------
>   17 |     ENV PATH="/opt/venv/bin:$PATH"
>   18 |     COPY . /build
>   19 | >>> RUN pip install /build/
>   20 |
>   21 |     FROM base
> --------------------
> ```

Hey boss, just taking a look at this `docker` stuff now 👀 

In other news: 

- I updated the CI as best I can sort of "blind" (no god mode to tinker in the actual runners or kick off tag builds to test this) so can fumble about with the changes if you let me know what's borked
- I updated the dev versioning which I think achieves what we're looking for (commit distance, etc.)
- I updated the docs - but these are super subjective so I'm happy to blast whatever you want in there

cyberw commented:
Build times are slightly increased. I guess it is kind of expected but could you have a quick look to see if it could be optimized? I guess adding the lock file would be the obvious thing (ideally we'd test with _and_ without a lock file, but the important thing is to speed things up - especially the "windows" and "linux" builds (the ones that just run the most basic tests)

cyberw commented:
> * I updated the CI as best I can sort of "blind" (no god mode to tinker in the actual runners or kick off tag builds to test this) so can fumble about with the changes if you let me know what's borked
> * I updated the dev versioning which I think achieves what we're looking for (commit distance, etc.)
> * I updated the docs - but these are super subjective so I'm happy to blast whatever you want in there
> * We're up to date with `master` again but this will likely need squashed on merge because the rebase is nasty

I think it is starting to look really good. I'll have a look at all the changes again tomorrow or the day after. I wont merge this until after our next release (but that should be very soon), so it can bounce around for a bit before we release it to everyone :)

cyberw commented:
Could be worth exploring something like this: https://github.com/marketplace/actions/python-poetry-action

mquinnfd commented:
> > * I updated the CI as best I can sort of "blind" (no god mode to tinker in the actual runners or kick off tag builds to test this) so can fumble about with the changes if you let me know what's borked
> > * I updated the dev versioning which I think achieves what we're looking for (commit distance, etc.)
> > * I updated the docs - but these are super subjective so I'm happy to blast whatever you want in there
> > * We're up to date with `master` again but this will likely need squashed on merge because the rebase is nasty
> 
> I think it is starting to look really good. I'll have a look at all the changes again tomorrow or the day after. I wont merge this until after our next release (but that should be very soon), so it can bounce around for a bit before we release it to everyone :)

That makes sense 👍 be nice to get eyeballs on it once it's fully polished

I've:
- Had a stab at fixing the `Dockerfile` - local builds seem to run OK now, entrypoint looks OK to me
- Added hidden development dependencies (`tox`, `ruff`, etc.) to an optional dev group in the `pyproject.toml` dependencies (things that are mentioned in the docs but not managed or pinned currently)
  - Could make this a default thing and omit from production builds
- Moved the docs dependencies to an optional dependency group, which helps ensure that there are never any version conflicts, these are only installed by specifying `install --with docs` for example

mquinnfd commented:
> Build times are slightly increased. I guess it is kind of expected but could you have a quick look to see if it could be optimized? I guess adding the lock file would be the obvious thing (ideally we'd test with _and_ without a lock file, but the important thing is to speed things up - especially the "windows" and "linux" builds (the ones that just run the most basic tests)

Yeah I think we have a little additional overhead for installing `poetry` which is inevitable, but I think caching would mitigate this for the most part, I see we have caching on for `pip`'s cache, which IIRC `poetry` doesn't use, I'm guessing I can just switch the `pip` caching step over to look at `poetry`'s cache instead

mquinnfd commented:
Note to self, WIP:
- Handle `poetry` caching in Github Actions (speed up repeated `tox` tests)
- Look at using `poetry` step in the Actions themselves: https://github.com/marketplace/actions/python-poetry-action
- Fix `ReadTheDocs` build step with the new dependency group
- Consider pinning `dev` dependency versions
- Include `dev` dependencies for local development steps

cyberw commented:
Docker works nicely for me now.

cyberw commented on pyproject.toml:
```diff
@@ -42,22 +33,126 @@ classifiers = [
+generate-setup-file = false
+
+[tool.poetry.urls]
+"Help/Questions" = "https://github.com/orgs/locustio/discussions/"
```
Can you make the first link to the docs?

mquinnfd replied:
The repo/docs URLs tend to be first-class citizens in the file, and are listed under the package itself (ln16)

I think these are mostly "other unspecified URLs" - would you want one of these to be docs also?

(sorry for RTD spam builds, think we're good now)

cyberw replied:
Ah. No that should be fine then.

cyberw commented:
If you get really tired battling GH actions you can try this to run them locally https://github.com/nektos/act

mquinnfd commented:
> If you get really tired battling GH actions you can try this to run them locally https://github.com/nektos/act

I'm mostly concerned I'm burning through any gratis CI runner minutes you may get for this project, so being able to run these locally for testing the changes would be great

mquinnfd commented:
Note to self, WIP:

- [ ] Fix caching in GH actions with poetry `venv`
- [ ] Fix `tox` builds with `poetry` on local GH runner (nektos/act)

cyberw commented:
> > If you get really tired battling GH actions you can try this to run them locally https://github.com/nektos/act
> 
> I'm mostly concerned I'm burning through any gratis CI runner minutes you may get for this project, so being able to run these locally for testing the changes would be great

I think it is fine, IIRC the only limit is on the number of concurrent runs, not total minutes.

cyberw commented:
Hi @mquinnfd ! Do you think you'll have time to finish up, or do you want me or @andrewbaldwin44 to take a look? Would love to see this merged soon :)

mquinnfd commented:
> Hi @mquinnfd ! Do you think you'll have time to finish up, or do you want me or @andrewbaldwin44 to take a look? Would love to see this merged soon :)

Hey boss 👋 
Sorry had a busy week 🫠 

Super happy if people wanna crack on with this as well
I'll try and carve out more time for it this week though

Was basically just running the GH Actions via `act` - then time to stand back and make sure it all looks OK

mquinnfd commented:
Was something like juggling, for the main `tox` steps, what has `poetry` installed and where, as well as having to build the web UI on a different runner stage (with `node` etc.) - 

I don't have a lot of experience with GH Actions tbh so been fumbling about with it but the local runner has been nice

cyberw commented:
Its not super urgent, just wanted to make sure you hadn't given up :)

mquinnfd commented:
> Its not super urgent, just wanted to make sure you hadn't given up :)

Not getting rid of me that easily 🚀

cyberw commented:
Windows is terrible. If you want me to debug it for you I think I have a machine around here somewhere :)

mquinnfd commented:
Hahah 👀 I actually have a Windows machine next to me, I'll give this a blast at the weekend on there just to debug the low-level process stuff that seems to be happening

I've got the other tests up and running but I need to tidy and optimise this then have a look at the packaging logic etc.

mquinnfd commented:
Finally got time to work on this on a Windows machine, seems like `poetry install` doesn't wanna add the editable install of `locust` to the path for some reason.

Can't find much about this online (not used Poetry on Windows at all) so easy solution is just specifically building and installing the wheel on the Windows runner specifically. This is because the fail-fast tests do call `locust` as a path binary, which is why this broke in the first place.

Not sure - there are faster ways to get the currently installed `locust` build on the path like just running `poetry run where locust` and adding that to the path, but this is arguably easier as it doesn't involve adding a `.bat` script somewhere. I could try doing this to shave the extra minute off the Windows test build time though.

For now at least it's all green.

It would make sense to test the `twine` publishing steps as well though, but I would have to maybe set this up on my fork to a private PyPi repo 🤔

cyberw commented on .vscode/settings.json:
```diff
@@ -1,39 +1,39 @@
-        "**/out.txt": true,
-        "**/.env": true
-    },
+    // "files.exclude": {
```
Was this an accidental change? Some of these might no longer be relevant, but I'm pretty sure they don't ALL deserve to be removed :)

mquinnfd replied:
Hey sorry yeah I'll revert this, I don't work like this locally as I like to see all the hidden files (versioning one, etc.) but clearly did a careless commit all here 😱

cyberw commented on Makefile:
```diff
@@ -1,23 +1,45 @@
+install: setup_dependencies
+	poetry install --with dev
+
+.ONESHELL:
```
Maybe add a comment about why this is needed (is it that poetry config sets env vars or somehthing?)

mquinnfd replied:
Actually I think this is probably unnecessary after removing the poetry installation from the `Makefile` - it was originally so that the `install poetry -> add plugin` worked end to end but it's not there anymore.

cyberw commented on Makefile:
```diff
@@ -1,23 +1,45 @@
 
+.SILENT:
+.PHONY: serve_docs
+serve_docs:
```
Does this work better than just browsing the files locally? (file://.../docs/_build/index.html)

mquinnfd replied:
I don't think it necessarily works _better_ - I just like knowing what an HTTP server will do to the files (this ends up being served by RTD)

Things like the logs _can_ be useful to see if anything is messed up (which reading an HTML file itself won't without looking at the browser console or something)

<img width="898" alt="image" src="https://github.com/locustio/locust/assets/82032454/fe67d571-7b6b-4a2e-9cf1-98fb033592ed">

That and it's just a fairly easy way to load the docs (we could even have the default browser pop open `http://localhost` so that it just starts showing you the docs you built).

It's really mostly a suggestion though, I can pap this out to slim down the file 👍

cyberw replied:
I think it make sense the way you did it. Maybe just add a comment about what it adds compared to browsing the files directly?

cyberw commented on Makefile:
```diff
@@ -1,23 +1,45 @@
+.SILENT:
+.PHONY: check-poetry
+check-poetry:
+	command -v poetry >/dev/null 2>&1 || { echo >&2 "Locust requires the poetry binary to be available in this shell to build the Python package.\nSee: https://docs.locust.io/en/stable/developing-locust.html#install-locust-for-development"; exit 1; }
```
I may have found a nicer way to do this:
`which poetry > /dev/null ||  {...`

mquinnfd replied:
Yeah I believe `which` is unix-specific - the windows equivalent would be `where` 

Come to think of it, I think `command` itself is POSIX compliant and I currently have no idea if this works on Windows either 😓 

Cross platform options likely boil down to:
- write a couple more lines (have cases for each platform)
- call out to the system's Python and have that check (we need Python anyway right)
- maybe just call the binary and catch the errors to the pipe anyway

cyberw commented on locust/contrib/fasthttp.py:
```diff
@@ -438,7 +438,7 @@ class FastResponse(CompatResponse):
     _response: HTTPSocketPoolResponse | None = None
 
-    encoding: str | None = None
+    encoding: str | float | None = None
```
When is encoding a float?

mquinnfd replied:
This is just one the type checks were balking at:

Technically using `chardet` 's `detect` method as a fallback to try and find out the encoding, the returned dictionary _values_ for the key _encoding_ (`Dict[str, str | float | None]`) can be `str | float | None` so it was complaining 🙃  

<img width="1137" alt="image" src="https://github.com/locustio/locust/assets/82032454/2eddc0ab-9d60-43e5-8ffd-383a55cdaec0">

mquinnfd replied:
It is probably super unlikely it's just technically a possibility at least in terms of permissable types

cyberw commented on pyproject.toml:
```diff
@@ -1,29 +1,20 @@
+license = "MIT"
+readme = "README.md"
+authors = ["Jonatan Heyman", "Carl Byström", "Joakim Hamrén", "Hugo Heyman"]
+maintainers = ["Lars Holmberg", "Jonatan Heyman"]
```
Please add Andrew Baldwin to this list.

cyberw commented:
Looking really good now. 

As for testing twine/publishing, its ok to break the publishing for a little while (maybe time the merge for when you know you have a couple hours to fix any issues that might arise?). I do have a pypy test account where we could publish, but meh.

I did have an issue with the docker image. Can you try it and see if the same thing happened to you?

```
> docker build -t locustio/locust:latest .
...
> docker run -p 8089:8089 -v $PWD:/mnt/locust locustio/locust -f /mnt/locust/locustfile.py
[2024-06-23 07:59:40,884] 474c796dc630/INFO/locust.main: Starting web interface at http://0.0.0.0:8089
[2024-06-23 07:59:40,890] 474c796dc630/INFO/locust.main: Starting Locust 2.29.0.dev110
(and then I opened the start page)
Traceback (most recent call last):
  File "/opt/venv/lib/python3.11/site-packages/flask/app.py", line 880, in full_dispatch_request
    rv = self.dispatch_request()
         ^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/venv/lib/python3.11/site-packages/flask/app.py", line 865, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/venv/lib/python3.11/site-packages/locust/web.py", line 565, in wrapper
    return view_func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/venv/lib/python3.11/site-packages/locust/web.py", line 170, in index
    return render_template("index.html", template_args=self.template_args)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/venv/lib/python3.11/site-packages/flask/templating.py", line 149, in render_template
    template = app.jinja_env.get_or_select_template(template_name_or_list)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/venv/lib/python3.11/site-packages/jinja2/environment.py", line 1084, in get_or_select_template
    return self.get_template(template_name_or_list, parent, globals)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/venv/lib/python3.11/site-packages/jinja2/environment.py", line 1013, in get_template
    return self._load_template(name, globals)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/venv/lib/python3.11/site-packages/jinja2/environment.py", line 972, in _load_template
    template = self.loader.load(self, name, self.make_globals(globals))
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/venv/lib/python3.11/site-packages/jinja2/loaders.py", line 126, in load
    source, filename, uptodate = self.get_source(environment, name)
                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/venv/lib/python3.11/site-packages/flask/templating.py", line 65, in get_source
    return self._get_source_fast(environment, template)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/venv/lib/python3.11/site-packages/flask/templating.py", line 99, in _get_source_fast
    raise TemplateNotFound(template)
jinja2.exceptions.TemplateNotFound: index.html

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/venv/lib/python3.11/site-packages/gevent/pywsgi.py", line 1107, in handle_one_response
    self.run_application()
  File "/opt/venv/lib/python3.11/site-packages/gevent/pywsgi.py", line 1053, in run_application
    self.result = self.application(self.environ, self.start_response)
                  ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/venv/lib/python3.11/site-packages/flask/app.py", line 1498, in __call__
    return self.wsgi_app(environ, start_response)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/venv/lib/python3.11/site-packages/flask/app.py", line 1476, in wsgi_app
    response = self.handle_exception(e)
               ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/venv/lib/python3.11/site-packages/flask_cors/extension.py", line 178, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
                                                ^^^^^^^^^^^^^^^^^^
  File "/opt/venv/lib/python3.11/site-packages/flask/app.py", line 1473, in wsgi_app
    response = self.full_dispatch_request()
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/venv/lib/python3.11/site-packages/flask/app.py", line 882, in full_dispatch_request
    rv = self.handle_user_exception(e)
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/venv/lib/python3.11/site-packages/flask_cors/extension.py", line 178, in wrapped_function
    return cors_after_request(app.make_response(f(*args, **kwargs)))
                                                ^^^^^^^^^^^^^^^^^^
  File "/opt/venv/lib/python3.11/site-packages/flask/app.py", line 772, in handle_user_exception
    return self.ensure_sync(handler)(e)  # type: ignore[no-any-return]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/venv/lib/python3.11/site-packages/locust/web.py", line 155, in handle_exception
    f"UI got request for {request.path}, but it resulted in a {error.code}: {error.name}",
                                                               ^^^^^^^^^^
AttributeError: 'TemplateNotFound' object has no attribute 'code'
2024-06-23T07:59:44Z {'REMOTE_ADDR': '::ffff:172.17.0.1', 'REMOTE_PORT': '54400', 'HTTP_HOST': '0.0.0.0:8089', (hidden keys: 25)} failed with AttributeError
```

mquinnfd commented:
> Looking really good now.
> 
> As for testing twine/publishing, its ok to break the publishing for a little while (maybe time the merge for when you know you have a couple hours to fix any issues that might arise?). I do have a pypy test account where we could publish, but meh.
> 
> ...
>

Yeah sounds good to me I can set aside some time for this 👍 
Might end up with some good suggestions for optimising the flow etc. if we make the PR for review at some point as well

mquinnfd commented:
Ok the docker image works now, thanks! I actually put the prebuilt web UI in the wrong location originally.

Updated version should work for you too. (sorry `http://example.com` whoever you are)

<img width="929" alt="image" src="https://github.com/locustio/locust/assets/82032454/a2c69dd7-b5ee-4f46-85cb-4406c4be2a0e">
<img width="1504" alt="image" src="https://github.com/locustio/locust/assets/82032454/654ce41b-dcbc-4d3a-a29d-8521dafcc7cc">

cyberw commented:
Great stuff! I’ll let @andrewbaldwin44 have a whirl when he gets back on Tuesday. Is there something left to do before you feel it is ready to merge? If not, you can take it out of draft and once Andrew has reviewed we’ll find a good time to merge.

andrewbaldwin44 commented on .github/workflows/tests.yml:
```diff
@@ -26,25 +26,45 @@ jobs:
+        with:
+          node-version: 18.x
+      - name: Install Yarn
+        run: npm install -g yarn
```
Think this should be fine but there's also https://github.com/marketplace/actions/github-action-for-yarn, which might be faster / more reliable

mquinnfd replied:
I think that group of Actions is for performing `yarn <thing>` commands mostly - I didn't see anything in there or the node installation Action about installing `yarn` itself 🤔 

This is because you need it available on the Runner's path, for when `yarn` is called in the shell (when running a build, while in `tox` or similar)

I did see [some discussion](https://github.com/actions/setup-node/issues/182) on this on the node Action but it's the usual descent into madness

andrewbaldwin44 replied:
Yeah than let's leave it as is, as I said I think this should be fine :)

andrewbaldwin44 commented:
I noticed when running the tests `python -m poetry run tox -e fail_fast_test_main` that it runs the build for the frontend as well. If it is only running the tests in `locust/test/test_main.py`, the build wouldn't be necessary. Maybe we could use `SKIP_PRE_BUILD`?

andrewbaldwin44 commented:
@mquinnfd I left a couple of small comments but overall this looks awesome :) Thanks so much for working on this and battling with Windows! :1st_place_medal:

mquinnfd commented on pyproject.toml:
```diff
@@ -1,29 +1,20 @@
-    "Flask-Cors >=3.0.10",
-    "pywin32; platform_system=='Windows'",
-]
+version = "2.29.0.dev29"
```
Absolutely not :D
This got brought in with a sloppy commit all

Fixed!

mquinnfd replied:
Absolutely not :D
This got brought in with a sloppy commit all

Fixed!

mquinnfd commented:
> @mquinnfd I left a couple of small comments but overall this looks awesome :) Thanks so much for working on this and battling with Windows! 🥇

Thanks for the review man 🙇 
I'll get these bits rounded up this week and then yeah might be time to open it up to the gallery, I'm sure there are some opinionated poetry bits and bobs that could be changed here

mquinnfd commented on locust/webui/.yarnrc:
```diff
@@ -0,0 +1 @@
+network-timeout 500000
```
Hey yeah - genuinely I started working on this on my old laptop (Macbook, intel, 16gb) and I needed this (despite having a 1gb down wired connection) otherwise I was getting the same timeouts locally that you'd see on an underspec'd docker container

On my shiny new laptop (Macbook, M3, 36gb) I doubt I'd need it - so I suppose it's a case of whether we should add it for lethargic build machines in the local build or not 🤔

andrewbaldwin44 replied:
Haha I've never seen yarn timeouts on my local before but I don't think having a longer timeout will hurt anyone :+1:

mquinnfd commented on pyproject.toml:
```diff
@@ -42,22 +33,140 @@ classifiers = [
+exclude = [
+    "locust/webui/public",
+    "locust/webui/src",
+    "locust/webui/*.*",
```
I think `locust/webui/**` seems to work OK yeah 👍 
I've added the `locust/webui/node_modules` as an explicit exclude as well - just in case 🙅

mquinnfd commented:
> I noticed when running the tests `python -m poetry run tox -e fail_fast_test_main` that it runs the build for the frontend as well. If it is only running the tests in `locust/test/test_main.py`, the build wouldn't be necessary. Maybe we could use `SKIP_PRE_BUILD`?

Hey this is a great call! This shaves some time off the Windows and Linux builds 👍 
Should see them skipping now - like below

```
Installing the current project: locust (2.29.1.dev122)
Preparing build environment with build-system requirements poetry-core>=1.0.0, poetry-dynamic-versioning>=1.0.0,<2.0.0
Skipping front end build...
fail_fast_test_main: commands[1]> poetry run python -m unittest -f locust.test.test_main

from tmp7ef0_06q_locustfile import TestUser1

.........ssssss.s...s.sss.......ss..s.....ss......sssss.
----------------------------------------------------------------------
Ran 56 tests in 135.491s

OK (skipped=21)
  fail_fast_test_main: OK (152.48=setup[1.67]+cmd[11.98,138.83] seconds)
  congratulations :) (152.62 seconds)
```

mquinnfd commented:
Going to open this for Public review and I believe the rest of these comments have been worked through - I'll take the opportunity while people eyeball it, to look at the publish steps next chance I get
